python/cudf/cudf/utils/queryutils.py

```diff
@@ -137,7 +137,7 @@ def query_compile(expr):
         key "args" is a sequence of name of the arguments.
     """
 
-    funcid = f"queryexpr_{np.uintp(hash(expr)):x}"
+    funcid = f"queryexpr_{np.uintp(abs(hash(expr))):x}"
```

Comment:
This is incorrect, `hash` returns in the semi-open interval `[-2**63, 2**63)`, but `abs` folds this to the closed interval `[0, 2**63]` (so you alias just shy of 50% of the values). Instead, you want to shift, I suspect, and then you don't need numpy in the loop at all:

```suggestion
    funcid = f"queryexpr_{hash(expr) + 2**63:x}"
```

That said, strings are hashable, so this seems like a weird way of constructing a cache key (it's somehow deliberately making it more likely that you get hash collisions and produce the wrong value).

I would have thought that this would do the trick:

```
@functools.cache
def query_compile(expr):
    name  = "queryexpr" # these are only looked up locally so names can collide
    info = query_parser(expr)
    fn = query_builder(info, name)
    args = info["args"]
    devicefn = cudf.jit(device=True)(fn)
    kernel = _wrap_query_expr(f"kernel_{name}", devicefn, args)
    info["kernel"] = kernel
    return info
```